w3sper.js/src/protocol-driver/mod.js

```diff
@@ -1035,27 +1035,113 @@ export const withdraw = async (info) =>
+ * @returns {Uint8Array|null} Converted Uint8Array or null for invalid input
+ */
+function strToArray(input) {
+    if (input == null || input === '') return null;
+    if (typeof input === 'string') {
+        return new TextEncoder().encode(input);
+    }
+    if (input instanceof ArrayBuffer) {
+        return input.byteLength ? new Uint8Array(input) : null;
+    }
+    if (input instanceof Uint8Array) {
+        return input.length ? input : null;
+    }
     return null;
-  }
+}
 
-  let buffer = null;
-  if (typeof memo === "string") {
-    buffer = new TextEncoder().encode(memo);
-  } else if (memo instanceof ArrayBuffer) {
-    buffer = new Uint8Array(memo);
-  } else if (memo instanceof Uint8Array) {
-    buffer = memo;
-  }
+/**
+ * Extracts pointers to a given buffer and its length
+ * @param ptr_obj - object in which pointers will be set
+ * @param data_ptr_name - name of data pointer attribute to be set
+ * @param len_ptr_name - name of data length pointer attribute to be set
+ * @param buffer - data to which pointers will be created
+ * @param malloc - memory allocating function
+ * @param memcpy - memory copying function
+ * @returns {Promise<void>}
+ */
+async function setPointers(ptr_obj, data_ptr_name, len_ptr_name, buffer, malloc, memcpy) {
+    ptr_obj[data_ptr_name] = await malloc(buffer.byteLength);
+    let len = buffer.byteLength;
+    await memcpy(ptr_obj[data_ptr_name], buffer);
+
+    const len_buf = new Uint8Array(4);
+    new DataView(len_buf.buffer).setUint32(0, len, true);
+    ptr_obj[len_ptr_name] = await malloc(4);
+    await memcpy(ptr_obj[len_ptr_name], len_buf);
+}
 
-  if (!buffer) {
-    return null;
-  }
+/**
+ * Serializes given payload to rkyv format
+ * @param payload - object containing payload, e.g. memo or contract call data
+ * @param malloc - memory allocating function
+ * @param memcpy - memory copying function
+ * @param create_tx_data - low level function performing the actual serialization
+ * @returns {Promise<Uint8Array<ArrayBuffer>|null>} Buffer containing serialized data
+ */
+async function serializePayload(payload = null, malloc, memcpy, create_tx_data) {
+    if (!payload) return null;
+    const ptr = Object.create(null);
+    let code;
+    let ret;
+
+    if ('memo' in payload && payload.memo) {
+        let buffer = strToArray(payload.memo);
+        if (!buffer) return null;
+        await setPointers(ptr, "memo", "memo_len", buffer, malloc, memcpy);
+        ret = await malloc(4);
+        ptr.dummy_contract_id = await malloc(32);
+        code = await create_tx_data(
+            null,
+            null,
+            null,
+            null,
+            ptr.dummy_contract_id,
+            ptr.memo_len,
+            ptr.memo,
+            ret
+        )
+    } else if ('fnName' in payload && 'fnArgs' in payload && 'contractId' in payload) {
+        // fn_name
+        let fn_name_arg = strToArray(payload.fnName);
```

Comment:
Should be `const` as it's never reassigned.

Reply:
done